doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,311 @@
+\ingroup module_hidden
+
+\page goto-program-transformations Goto Program Transformations
+
+\section required-transforms Core Transformation Passes
+
+This section lists the transformation passes that must have been applied for the
+goto model to be in core goto format.
+
+Note that the passes are listed below in the order they are currently called in
+CBMC. While all dependencies on the ordering are not fully documented, the
+following order should be used.
+
+\subsection assembly-transform Removal/Lowering of Assembly
+
+This pass goes through the goto model and removes or lowers instances of
+assembly intructions. Assembly instructions are stored in instances of the
+`other` instruction.
+
+The implementation of this pass is called via the \ref remove_asm(goto_modelt &)
+function. For full documentation of this pass see \ref remove_asm.h
+
+<em>This pass has no predecessor.</em>
+
+\subsection linking-transform Linking of Standard Libraries
+
+This pass links the function definitions from the Cprover standard library
+models to the current goto model.
+
+The implementation of this pass is called via the \ref
+link_to_library(goto_modelt &, message_handlert &, const std::function<void(const std::set<irep_idt> &, const symbol_tablet &, symbol_tablet &, message_handlert &)> &) "link_to_library function."
+
+<em>Predecessor pass is \ref assembly-transform .</em>
+
+\subsection function-pointer-transform Removal/Lowering of Function Pointers
+
+This pass finds all the instructions which are function calls to the value of a
+function pointer. Each instruction is then replaced with a switch block. The
+switch block contains a case for each of the potential targets of the  function
+pointer. The targets are found by looking for all functions in the goto program
+that approximately match the type of the function pointer.
+
+Note that for this pass to work as intended, all potential targets of calls to
+function pointers must be in the model. Otherwise they will not be added to the
+relevant switch blocks. This may cause the assertion that the switch blocks'
+fallthrough is unreachable to be shown to be violated via an invalid counter
```

Comment:
:thinking: The counter example would be correct in terms of the model, but incorrect compared to a concrete execution. The counter example states a way of reaching a fall though which does not exist in the source program. So a concrete execution matching the trace would result in a real function (which was removed from the model) being called instead of the generated fall-through.